src/pure/RealTimeQueue.mo

```diff
@@ -0,0 +1,1026 @@
+///
+/// Note that some operations that traverse the elements of the queue (e.g. `forEach`, `values`) preserve the order of the elements,
+/// whereas others (e.g. `map`, `contains`, `any`) do NOT guarantee that the elements are visited in any order.
+/// The order is undefined to avoid allocations, making these operations more efficient.
+///
+/// ```motoko name=import
+/// import Queue "mo:base/pure/RealTimeQueue";
+/// ```
+
+import Types "../Types";
+import List "List";
+import Option "../Option";
+import { trap } "../Runtime";
+import Iter "../Iter";
+
+module {
+  /// The real-time queue data structure can be in one of the following states:
+  ///
+  /// - `#empty`: the queue is empty
+  /// - `#one`: the queue contains a single element
+  /// - `#two`: the queue contains two elements
+  /// - `#three`: the queue contains three elements
+  /// - `#idles`: the queue is in the idle state, where `l` and `r` are non-empty stacks of elements fulfilling the size invariant
+  /// - `#rebal`: the queue is in the rebalancing state
+  public type Queue<T> = {
+    #empty;
+    #one : T;
+    #two : (T, T);
+    #three : (T, T, T);
+    #idles : (Idle<T>, Idle<T>);
+    #rebal : States<T>
+  };
+
+  /// Create a new empty queue.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// persistent actor {
+  ///   let queue = Queue.empty<Nat>();
+  ///   assert Queue.isEmpty(queue);
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(1)`.
+  ///
+  /// Space: `O(1)`.
+  public func empty<T>() : Queue<T> = #empty;
+
+  /// Determine whether a queue is empty.
+  /// Returns true if `queue` is empty, otherwise `false`.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// persistent actor {
+  ///   let queue = Queue.empty<Nat>();
+  ///   assert Queue.isEmpty(queue);
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(1)`.
+  ///
+  /// Space: `O(1)`.
+  public func isEmpty<T>(queue : Queue<T>) : Bool = switch queue {
+    case (#empty) true;
+    case _ false
+  };
+
+  /// Create a new queue comprising a single element.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// persistent actor {
+  ///   let queue = Queue.singleton<Nat>(25);
+  ///   assert Queue.size(queue) == 1;
+  ///   assert Queue.peekFront(queue) == ?25;
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(1)`.
+  ///
+  /// Space: `O(1)`.
+  public func singleton<T>(element : T) : Queue<T> = #one(element);
+
+  /// Determine the number of elements contained in a queue.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// persistent actor {
+  ///   let queue = Queue.singleton<Nat>(42);
+  ///   assert Queue.size(queue) == 1;
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(1)`.
+  ///
+  /// Space: `O(1)`.
+  public func size<T>(queue : Queue<T>) : Nat = switch queue {
+    case (#empty) 0;
+    case (#one(_)) 1;
+    case (#two(_, _)) 2;
+    case (#three(_, _, _)) 3;
+    case (#idles((l, nL), (r, nR))) {
+      debug assert Stacks.size(l) == nL and Stacks.size(r) == nR;
```

Comment:
```suggestion
      debug assert Stacks.size(l) == nL and Stacks.size(r) == nR;
```
I wonder if we should disable this assert since peope are probably all building with debug on.

Reply:
Maybe you could check it in the tests instead?